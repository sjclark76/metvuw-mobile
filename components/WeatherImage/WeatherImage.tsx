import LazyLoad from 'react-lazyload'

export interface WeatherImageProps {
  imageSrc: string
  imageAlt: string
  isRainForecast: boolean
}
const WeatherImage = (props: WeatherImageProps) => {
  const rainStyle = {
    objectPosition: '0% 62%',
  }

  const satelliteStyle = {
    objectPosition: '0% 0%',
  }

  return (
    <div
      data-testid="weather-image"
      className="inline-block max-w-full overflow-hidden relative box-border m-0"
    >
      <div className="box-border block, max-w-full">
        <img
          className="max-w-full block m-0 border-none p-0"
          alt=""
          aria-hidden="true"
          role="presentation"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNzExIiBoZWlnaHQ9IjQ5MiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
        />
      </div>
      <LazyLoad height={200} offset={600}>
        <img
          alt={props.imageAlt}
          src={props.imageSrc}
          decoding="async"
          className="absolute top-0 left-0 bottom-0 right-0, box-border p-0 border-none m-auto block w-0, h-0 min-w-full max-w-full min-h-full max-h-full object-cover"
          style={props.isRainForecast ? rainStyle : satelliteStyle}
        />
      </LazyLoad>
    </div>
  )
}
export default WeatherImage
