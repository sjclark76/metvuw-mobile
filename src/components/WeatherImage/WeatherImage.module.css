.aspectRatioBox {
    width: 100%;
    height: 0;
    padding-top: 64%;
    position: relative;
    overflow: hidden;
}

.croppedImage {
    position: absolute;
    top: 0;
    width: 100%;
    height: 100%;
    object-fit: cover;
    object-position: 50% 50%;
    transform:  scale(1.29) translateY(-12px) translate(0, 0) rotate(0deg);
}