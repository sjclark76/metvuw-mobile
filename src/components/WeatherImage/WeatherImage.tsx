import { LazyLoadImage } from 'react-lazy-load-image-component'

import {
  radarImageDimensions,
  rainImageDimensions,
  satelliteImageDimensions,
  upperAirImageDimensions,
} from '@/shared/helpers/v2/imageCompression/imageDimensions'
import { ChartType } from '@/shared/types/ChartType'

export interface WeatherImageProps {
  chartType: ChartType
  imageAlt: string
  imageSrc: string | undefined
  isLazy: boolean
}

function extraImageAttribute(
  chartType: ChartType,
  imageSrc: string | undefined,
) {
  const smallImageSrc = imageSrc?.replace('images', 'small-images')

  switch (chartType) {
    case 'Satellite':
      return {
        srcSet: `${smallImageSrc} 300w, ${imageSrc} 840w`,
        sizes: '(min-width: 900px) 840px, calc(93.1vw + 21px)',
        ...satelliteImageDimensions,
      }
    case 'Upper Air':
      return {
        srcSet: `${smallImageSrc} 300w, ${imageSrc} 760w`,
        sizes: '(min-width: 820px) 760px, calc(92vw + 24px)',
        ...upperAirImageDimensions,
      }
    case 'Radar':
      return {
        ...radarImageDimensions,
      }
    case 'Rain':
      return {
        ...rainImageDimensions,
      }
    default:
      return {}
  }
}

const WeatherImage = ({
  imageAlt,
  imageSrc,
  chartType,
  isLazy,
}: WeatherImageProps) => {
  const attributes = extraImageAttribute(chartType, imageSrc)

  if (!imageSrc) {
    return (
      <div data-testid="weather-image-placeholder">Image not available</div>
    )
  }

  const commonProps = {
    src: imageSrc,
    ...attributes,
    crossOrigin: 'anonymous',
    style: {
      display: 'block', // Make image a block element
      width: '100%', // Make image attempt to fill container width
      maxWidth: '100%', // Ensure it doesn't overflow container
      height: 'auto', // Maintain aspect ratio
      ...(attributes &&
      typeof attributes === 'object' &&
      'style' in attributes &&
      typeof attributes.style === 'object'
        ? attributes.style
        : {}),
    },
  }

  return (
    // Ensure this div takes full width of its parent, constraining the image
    <div data-testid="weather-image" className="w-full">
      {isLazy ? (
        <LazyLoadImage alt={imageAlt} {...commonProps} threshold={200} />
      ) : (
        // eslint-disable-next-line jsx-a11y/alt-text
        <img alt={imageAlt} {...commonProps} fetchPriority="high" />
      )}
    </div>
  )
}
export default WeatherImage
