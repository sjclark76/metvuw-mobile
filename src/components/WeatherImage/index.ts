import WeatherImage from './WeatherImage'
export default WeatherImage
