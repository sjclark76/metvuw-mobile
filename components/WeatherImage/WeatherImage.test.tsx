import { render } from '@testing-library/react'
import WeatherImage from './WeatherImage'

describe('Weather Image', () => {
  test('it should render correctly', () => {
    const { container } = render(
      <WeatherImage isRainForecast imageAlt="foo" imageSrc="baa" />
    )

    expect(container).toMatchSnapshot()
  })
})
