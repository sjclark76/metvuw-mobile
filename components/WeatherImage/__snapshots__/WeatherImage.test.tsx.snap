// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Image it should render correctly 1`] = `
<div>
  <div
    class="aspectRatioBox"
    data-testid="weather-image"
  >
    <img
      alt="foo"
      class="croppedImage"
      src="baa"
      style="object-position: 0% 62%; min-height: 85%;"
    />
  </div>
</div>
`;
